server/src/main/java/org/elasticsearch/search/rank/rerank/RerankingQueryPhaseRankShardContext.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.rank.rerank;
+
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.elasticsearch.search.rank.RankShardResult;
+import org.elasticsearch.search.rank.context.QueryPhaseRankShardContext;
+import org.elasticsearch.search.rank.feature.RankFeatureDoc;
+import org.elasticsearch.search.rank.feature.RankFeatureShardResult;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The {@link RerankingQueryPhaseRankShardContext} is responsible for combining the different shard-level query results, and
+ * then pack them to a {@code RankFeatureShardResult} to return to the coordinator. If a document is found in more than one queries, we
+ * only keep the max score for that document. This is to be treated with care, as different queries might have different score ranges that
+ * could affect the final ranking.
+ */
+public class RerankingQueryPhaseRankShardContext extends QueryPhaseRankShardContext {
+
+    public RerankingQueryPhaseRankShardContext(List<Query> queries, int windowSize) {
+        super(queries, windowSize);
+    }
+
+    @Override
+    public RankShardResult combineQueryPhaseResults(List<TopDocs> rankResults) {
```

Comment:
One of the main features of this new phase is the ability to rerank a result set return from RRF. How does this work if we merge the various score docs? Or does this run after rrf?

Reply:
Currently, combining different `RankBuilder` implementations (e.g. rrf and semantic reranker) is not an options (restricted also by the parser-only approach). This is something to be addressed in subsequent PRs.